### PR TITLE
feat(apollo-react): unified CanvasEdge with waypoint editing and pluggable routing

### PR DESCRIPTION
**Pull request by snuziale**
## Summary

Introduces **`CanvasEdge`** — a single canvas edge component whose behaviors compose via flags on `data` — and refactors **`SequenceEdge`** into a thin, backward-compatible preset on top of it (`routing: 'handle'`, execution + toolbar enabled, legacy `hideToolbar` translated).

### What's new

- **Waypoint routing** (default): orthogonal segments with rounded corners, hover-revealed segment drag handles, double-click to insert/remove waypoints, grid snapping, and waypoint rebalancing while a connected node is dragged. Controlled mode via `onWaypointsChange` for undo/redo or server-side persistence.
- **Shared edge layer**: `EdgePath` / `EdgeArrow` / `EdgeLabel` / handle primitives, geometry + waypoint math (unit-tested), color resolution, and memo comparator extracted from the old SequenceEdge monolith.
- **Pluggable routing**: an `EdgeRouter` contract + `useGraphRouter` orchestration hook. Subscribes via a geometry/connectivity fingerprint, so selection changes and the router's own writes don't re-trigger routing; manual waypoints always take priority over routed output.
- **`animated` support** carried through from main (no `strokeDasharray` override so React Flow's dashdraw CSS applies).
- **Stories**: editing, non-editable, handle routing (the SequenceEdge preset), controlled mode, pluggable router, and execution states (Start → Step → End chains where node + edges share one mocked status, with editing enabled to show behaviors compose).

https://github.com/user-attachments/assets/3b2393e2-75ab-46a7-9362-5ec4f58d9475

### Backward compatibility (SequenceEdge consumers)

API- and behavior-compatible: same props/`data` contract (incl. `hideToolbar`), same path geometry (incl. self-loop/`loopBack`), same execution colors/animations, same toolbar conditions, perf-neutral. Two deliberate visual deltas for reviewers to confirm:

1. **Selected edges render 3px stroke** (was 2px) when no explicit `style.strokeWidth` is set.
2. **`style.strokeDasharray` is no longer honored on diff-removed edges** — they always use the standard `5,5` dash.

Both will show up in visual regression baselines; flag if either should be reverted instead.

## Test plan

- [x] `vitest` canvas suite: 75 files / 1487 tests passing (includes new geometry + waypoint tests)
- [x] `tsc --noEmit` and Biome clean
- [ ] Visual regression run to consciously accept the two deltas above
- [ ] Manual pass over the new Storybook stories (Components/Edges/CanvasEdge)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: packages/apollo-react/src/canvas/components/Edges/shared/primitives/WaypointHandle.tsx [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+  const visible = forceVisible || isHovered;
+
+  return (
+    <g
+      transform={`translate(${waypoint.x}, ${waypoint.y})`}
+      onMouseEnter={() => setIsHovered(true)}
+      onMouseLeave={() => setIsHovered(false)}
+      onMouseDown={(e) => handlers.onMouseDown(index, e)}
+      onDoubleClick={() => handlers.onDoubleClick(index)}
+      style={{ cursor: 'move' }}
```
[the remark this answers is not in the document]
Comment (snuziale) [reply packages/apollo-react/src/canvas/components/Edges/shared/primitives/WaypointHandle.tsx:39]: Fixed in 81c6cd0e — with one correction: a React-level `stopPropagation` can't actually prevent the zoom, because d3-zoom's `dblclick.zoom` listener sits on the renderer div and fires during native bubbling, before React's root-delegated synthetic dispatch. The handle now has the `nopan` class, which xyflow's zoom filter checks (`isWrappedWithClass`) and rejects at the source. `stopPropagation`/`preventDefault` were also added to keep React Flow's edge-level double-click handlers (selection, `onEdgeDoubleClick`) out of waypoint removal.

File: packages/apollo-react/src/canvas/components/Edges/shared/hooks/useWaypointEditor.ts [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+      onDoubleClick: (segmentIndex: number, e: React.MouseEvent) => {
+        if (!enabled) return;
+        // Convert screen coords to canvas/flow coords so the new waypoint
+        // lands at the cursor regardless of pan/zoom.
+        const flowPoint = screenToFlowPosition({ x: e.clientX, y: e.clientY });
+        const snapped = snapPointToGrid(flowPoint);
+        updateWaypoints(
+          insertWaypoint(pathPointsRef.current, waypointsRef.current, segmentIndex, snapped)
+        );
+      },
```
[the remark this answers is not in the document]
Comment (snuziale) [reply packages/apollo-react/src/canvas/components/Edges/shared/hooks/useWaypointEditor.ts:208]: Fixed in 81c6cd0e. Same caveat as the waypoint handle: `stopPropagation` alone can't block d3's double-click zoom (its listener fires below React's delegation root), so the segment interaction line now carries the `nopan` class, and the editor's insert handler additionally stops propagation/default to keep edge-level double-click handlers quiet.

File: packages/apollo-react/src/canvas/components/Edges/shared/waypoints.ts [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+export function materializePathWaypoints(vertices: PathVertex[], stored: Waypoint[]): Waypoint[] {
+  return vertices.slice(1, -1).map((vertex) => {
+    const existing = vertex.waypointIndex >= 0 ? stored[vertex.waypointIndex] : undefined;
+    return existing ?? { id: generateWaypointId(), x: vertex.x, y: vertex.y };
+  });
+}
+
+export function generateWaypointId(): string {
+  return `wp-${Date.now()}-${Math.random().toString(36).substring(2, 9)}`;
+}
```
[the remark this answers is not in the document]
Comment (snuziale) [reply packages/apollo-react/src/canvas/components/Edges/shared/waypoints.ts:24]: Fixed in 81c6cd0e — `generateWaypointId` now uses `crypto.randomUUID()`, consistent with existing canvas usage (`mermaid-parser.ts`).

File: packages/apollo-react/src/canvas/components/Edges/shared/routing/defaultEdgeRouter.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+/**
+ * Per-edge naive router — independent orthogonal routing for each edge with
+ * no obstacle awareness. Equivalent to CanvasEdge's built-in fallback when no
+ * waypoints exist. Useful as a baseline implementation, a starting point for
+ * custom routers, or to satisfy the EdgeRouter contract without external deps.
+ *
+ * IDs are deterministic (`${edgeId}-r${i}`) so repeated calls with identical
+ * positions produce reference-identical waypoints — useful for memoization
+ * and predictable test assertions.
```
[the remark this answers is not in the document]
Comment (snuziale) [reply packages/apollo-react/src/canvas/components/Edges/shared/routing/defaultEdgeRouter.ts]: Fixed in 81c6cd0e — doc reworded: deterministic ids give *value-equal* waypoints across calls (fresh objects each time), which is what the `waypointsPositionallyEqual` write-guard and test assertions rely on; no claim of referential equality.

File: packages/apollo-react/src/canvas/components/Flow.stories.tsx [hunk 5]
```diff
@@ -164,7 +176,15 @@ const additionalNodeTypes = {
...
 };
 
-function DefaultStory({ useSmartHandles }: FlowStoryArgs) {
+// Preset wrapper following the SequenceEdge pattern: editing is enabled at
+// render time for every edge of this type, so no per-edge store state needs
+// to be patched (and newly connected edges pick it up automatically).
+function EditableCanvasEdge(props: CanvasEdgeProps) {
+  const data = useMemo(() => ({ enableEditing: true, ...props.data }), [props.data]);
+  return <CanvasEdge {...props} data={data} />;
+}
```
[the remark this answers is not in the document]
Comment (snuziale) [reply packages/apollo-react/src/canvas/components/Flow.stories.tsx]: Comment reworded in 81c6cd0e. The spread order is intentional — defaults-first/consumer-override-last is the same layering the `SequenceEdge` preset uses, and no edge in this story sets `enableEditing: false`. The comment now says editing *defaults on* rather than implying it's forced.

File: packages/apollo-react/src/canvas/components/Edges/SequenceEdge.tsx [hunk 6]
```diff
@@ -1,264 +1,29 @@
...
-// Offset to position start/arrow slightly away from the node edge
-const ARROW_OFFSETS: Record<Position, { x: number; y: number }> = {
-  [Position.Left]: { x: 8, y: 0 },
-  [Position.Right]: { x: -8, y: 0 },
-  [Position.Top]: { x: 0, y: 8 },
-  [Position.Bottom]: { x: 0, y: -8 },
+import { useMemo } from 'react';
+import { CanvasEdge } from './CanvasEdge';
+import type { CanvasEdgeData, CanvasEdgeProps } from './shared/types';
+
```
[the remark this answers is not in the document]
Comment (snuziale) [reply packages/apollo-react/src/canvas/components/Edges/SequenceEdge.tsx]: Restored in 81c6cd0e — `SequenceEdge` is again wrapped in `memo(…, areEdgePropsEqual)`. One clarification for the record: the un-memoized wrapper did not re-run on *every* store update — only when React Flow's EdgeWrapper re-rendered it (position/selection changes), and `CanvasEdge`'s memo uses the identical comparator, so the expensive render was already skipped. The delta was one wrapper call per jitter tick; restoring the outer memo eliminates even that and keeps the preset's perf semantics byte-identical to the old component.

File: packages/apollo-react/src/canvas/components/Edges/SequenceEdge.tsx [hunk 7]
```diff
@@ -1,264 +1,29 @@
...
+    return {
+      routing: 'handle',
+      enableExecution: true,
+      enableToolbar: hideToolbar !== true,
+      ...rest,
+    };
+  }, [props.data]);
+
+  return <CanvasEdge {...props} data={data} />;
 }
```
[the remark this answers is not in the document]
Comment (snuziale) [reply packages/apollo-react/src/canvas/components/Edges/SequenceEdge.tsx]: Done in 81c6cd0e — `memo(…, areEdgePropsEqual)` restored on the preset, legacy `hideToolbar` translation unchanged.

File: packages/apollo-react/src/canvas/components/Edges/shared/primitives/SegmentDragHandle.tsx [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+  horizontal: 'ns-resize',
+  vertical: 'ew-resize',
+};
+
+const HANDLE_RECT: Record<
+  SegmentOrientation,
+  { x: number; y: number; width: number; height: number }
+> = {
+  horizontal: { x: -12, y: -5, width: 24, height: 10 },
+  vertical: { x: -5, y: -12, width: 10, height: 24 },
```
Comment (CalinaCristian) [inline packages/apollo-react/src/canvas/components/Edges/shared/primitives/SegmentDragHandle.tsx:19]: NIT: maybe some consts for this?

Comment (snuziale) [reply packages/apollo-react/src/canvas/components/Edges/shared/primitives/SegmentDragHandle.tsx:19]: this whole object is the const :)

File: packages/apollo-react/src/canvas/components/Edges/CanvasEdge.tsx [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+  const storedWaypoints = data?.waypoints ?? EMPTY_WAYPOINTS;
+  const strokeStyle = data?.strokeStyle ?? 'solid';
+  const hideArrowHead = !!data?.hideArrowHead;
+  const label = data?.label;
+
+  const editingEnabled = !!data?.enableEditing && routing === 'waypoint' && !isReadOnly;
+
+  // While a connected node is dragged, the waypoints follow it (and persist on
+  // release); otherwise these are the stored waypoints unchanged.
+  const waypoints = useNodeDragRebalance({
```
Comment (CalinaCristian) [inline packages/apollo-react/src/canvas/components/Edges/CanvasEdge.tsx:67]: ### P2 - Disabled behaviors still install per-edge hook work

`enableEditing`, `enableExecution`, and `enableToolbar` read like hard gates, but this component still calls the editing, execution, and toolbar hooks for every edge. In particular, `useNodeDragRebalance` subscribes to the React Flow store even when `editingEnabled` is false, and `SequenceEdge` always takes that disabled editing path because it forces `routing: 'handle'`. On large canvases this adds per-edge subscription/context churn to the backward-compatible path.

**Fix:** split the optional behaviors into mounted-only child components/hooks, or add enabled-only wrapper hooks, so disabled editing/execution/toolbar paths do not subscribe or allocate beyond the visual edge render.

File: packages/apollo-react/src/canvas/components/Edges/shared/routing/useGraphRouter.ts [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+
+    let cancelled = false;
+    const result = router.route(request);
+
+    const handle = (routes: Awaited<ReturnType<EdgeRouter['route']>>) => {
+      if (cancelled) return;
+      const byId = new Map(routes.map((r) => [r.edgeId, r]));
+      setEdges((current) => {
+        let changed = false;
+        const next = current.map((edge) => {
```
Comment (CalinaCristian) [inline packages/apollo-react/src/canvas/components/Edges/shared/routing/useGraphRouter.ts]: ### P2 - Router writes metadata into every returned edge

`useGraphRouter` builds a request from every edge and writes `data.routedWaypoints` back into every edge id returned by the router, with no check that the edge is a waypoint `CanvasEdge` or that it is still router-controlled. In mixed graphs, a consumer router can dirty custom/legacy edge data and pollute undo or persistence state with render-only routing metadata.

**Fix:** filter the request/write path with a routeable-edge predicate, at minimum only waypoint `CanvasEdge` edges without manual `data.waypoints`; also clear stale `routedWaypoints` when an edge becomes manual or no longer routeable.

File: packages/apollo-react/src/canvas/components/Edges/SequenceEdge.tsx [hunk 11]
```diff
@@ -1,264 +1,30 @@
...
-      )}
-    </>
-  );
+/**
+ * Workflow edge preset — `CanvasEdge` with `routing: 'handle'`, execution and
+ * toolbar enabled by default. Consumer-supplied `data` overrides any default.
+ * Prefer `CanvasEdge` directly for new code; kept for backward compatibility
+ * with existing `type: 'sequence'` registrations.
+ */
+export const SequenceEdge = memo(function SequenceEdge(props: CanvasEdgeProps) {
```
Comment (CalinaCristian) [inline packages/apollo-react/src/canvas/components/Edges/SequenceEdge.tsx:18]: ### P2 - Compatibility claim needs component coverage

`SequenceEdge` was replaced by this preset wrapper, but the added tests only cover geometry and waypoint helpers. The legacy behavior is now distributed through `CanvasEdge`, `useEdgeGeometry`, `EdgePath`, `resolveEdgeColor`, and the wrapper's `hideToolbar` translation, so a regression in rendered behavior can pass the current unit tests.

**Fix:** add component or visual tests for the preserved legacy states: `hideToolbar`, execution/validation color priority, labels, `hideArrowHead`, diff added/removed styling, animated dash behavior, and the wrapper's default flag translation.

**Review by CalinaCristian — COMMENTED** [review]
Review posted from local pass. No P0/P1 blockers found; these are three P2 issues around performance, data-boundary behavior, and compatibility coverage. I also verified `pnpm --dir packages/apollo-react test` passes, `tsc --noEmit` passes after dependency packages are built, changed edge files lint clean, and CI checks are green.

**Review by CalinaCristian — APPROVED** [review]

File: packages/apollo-react/src/canvas/components/Edges/shared/routing/useGraphRouter.ts (hunk 10 above)
Comment (snuziale) [reply packages/apollo-react/src/canvas/components/Edges/shared/routing/useGraphRouter.ts]: Addressed in 2172a150. `useGraphRouter` now only routes and writes edges under router control: a new exported `defaultIsRoutable` predicate excludes handle-routed presets and edges with manual `data.waypoints`, and when an edge leaves router control its stale `routedWaypoints` key is removed (not just ignored), so render-only metadata can't reach persistence or undo state. The takeover transition is folded into the store fingerprint so the cleanup pass fires exactly on that frame — including when the last routable edge becomes manual.

One disclosed limit: the default predicate can't verify an edge *is* a CanvasEdge (type→component registration is consumer-defined), so a custom edge whose data lacks `routing`/`waypoints` defaults to routable. For mixed graphs the predicate is overridable and documented as composable: `useGraphRouter(router, (e) => defaultIsRoutable(e) && isMyEdge(e))`.

All of it is covered by a new `useGraphRouter.test.tsx` against the real React Flow store: request filtering, takeover cleanup, no write-echo re-routing, and zero-routable graphs never invoking the router.

File: packages/apollo-react/src/canvas/components/Edges/SequenceEdge.tsx (hunk 11 above)
Comment (snuziale) [reply packages/apollo-react/src/canvas/components/Edges/SequenceEdge.tsx:18]: Addressed in 2172a150 — added unit coverage locking the legacy contract: `resolveEdgeColor` priority matrix (diff > invalid > preview/selected > hover > status > default), `areEdgePropsEqual` (incl. `animated` and sub-pixel jitter suppression), `EdgePath` (solid/dashed/animated dash behavior, stroke widths, selection outline, opacity, interaction layer, readonly cursor — asserted as literals on purpose so constant retuning fails loudly), `EdgeLabel` and `EdgeArrow` rendering, the SequenceEdge preset's `hideToolbar` → `enableToolbar` translation and default flags, plus the router contract itself.

One sliver intentionally left to the visual regression suite (already on the PR test plan): the `previewEdge || isDiffRemoved → dashed` mapping lives inside `CanvasEdge`, which needs the full provider stack (canvas mode, execution/validation contexts, toolbar manifest) to render in a unit test — the diff *colors* are locked in resolveEdgeColor tests.

File: packages/apollo-react/src/canvas/components/Edges/CanvasEdge.tsx (hunk 9 above)
Comment (snuziale) [reply packages/apollo-react/src/canvas/components/Edges/CanvasEdge.tsx:67]: Agreed this is real, deferring to a dedicated follow-up PR rather than patching here — the proper fix inverts data flow (rebalanced waypoints feed geometry; execution color feeds the parent's path), which means splitting CanvasEdge by routing mode into conditionally-mounted behavior components. That same restructure retires the remaining altitude debt (`useEdgeGeometry` feeding dummy inputs to `useEdgePath`, the single-consumer `routing: 'handle'` branch).

For scale context on the backward-compatible path: the only *new* per-edge cost vs. the old SequenceEdge is one no-op store subscription (`useNodeDragRebalance`'s selector short-circuits on `enabled` to a boolean test); the toolbar and execution hooks ran unconditionally in the old component too.

File: packages/apollo-react/src/canvas/components/Edges/shared/routing/useGraphRouter.ts [hunk 12]
```diff
@@ -0,0 +1,223 @@
...
+ * The manual-waypoints clause is load-bearing: it is what makes the router
+ * yield to `useWaypointEditor` (which materializes routed points into
+ * `data.waypoints` on the first segment drag) and what arms the stale
+ * `routedWaypoints` cleanup. Custom predicates should compose with it rather
+ * than replace it: `(edge) => defaultIsRoutable(edge) && isMyEdge(edge)`.
+ */
+export function defaultIsRoutable(edge: Edge): boolean {
+  const data = edge.data as CanvasEdgeData | undefined;
+  return data?.routing !== 'handle' && !data?.waypoints?.length;
+}
```
[the remark this answers is not in the document]
Comment (snuziale) [reply packages/apollo-react/src/canvas/components/Edges/shared/routing/useGraphRouter.ts:42]: Good catch — confirmed: the preset applies `routing: 'handle'` at render time only, so SequenceEdge store edges were indistinguishable from waypoint edges and the previous default would have written `routedWaypoints` into them. Fixed in cd765896 by taking the explicit opt-in route you suggested: `defaultIsRoutable` now requires `data.routing === 'waypoint'`, so undeclared edges are never router-fed or polluted. The failure modes aren't symmetric — opt-out fails silently (persistence pollution), opt-in fails visibly (edges don't route until declared) — so opt-in is the safer default for a published API.

Note the deliberate asymmetry, documented on the predicate, the hook, and the `EdgeRouting` type: CanvasEdge still *renders* waypoint routing when `routing` is unset; only the router feed requires the explicit declaration. The PluggableRouter story now declares it, and the router test suite includes a store-level SequenceEdge shape (no `routing` field) asserting it's never routed or written to.

**Review by CalinaCristian — APPROVED** [review]

**Review by CalinaCristian — APPROVED** [review]
